signac/core/file_buffered_synced_collection.py

```diff
@@ -0,0 +1,283 @@
+# All rights reserved.
+# This software is licensed under the BSD 3-Clause License.
+"""Implement the FileBufferedSyncedCollection class.
+
+The FileBufferedSyncedCollection is a concrete implementation of the buffer
+protocol established by the BufferedSyncedCollection class. It uses an
+in-memory cache to store data when in buffered mode. It is suitable for
+use with any file-based backend because it performs integrity checks based on
+whether or not the underlying file has been modified while buffering was
+activated.
+"""
+
+import errno
+import hashlib
+import json
+import os
+import sys
+
+from typing import Any, Dict
+
+from .buffered_synced_collection import BufferedCollection
+from .errors import MetadataError
+
+
+def get_buffer_capacity():
+    """Get the current buffer capacity."""
+    return FileBufferedCollection._BUFFER_CAPACITY
+
+
+def set_buffer_capacity(new_capacity):
+    """Update the buffer capacity."""
+    FileBufferedCollection._BUFFER_CAPACITY = new_capacity
+
+
+def get_current_buffer_size():
+    """Get the total amount of data currently stored in the buffer."""
+    return FileBufferedCollection._CURRENT_BUFFER_SIZE
+
+
+class FileBufferedCollection(BufferedCollection):
+    """Implement buffering for SyncedCollections with file-based backends.
+
+    All file-based backends can use the same set of integrity checks prior to a
+    buffer flush. This class standardizes that protocol.
+    """
+
+    _cache: Dict[str, Any] = {}
+    _cached_collections: Dict[int, BufferedCollection] = {}
+    _BUFFER_CAPACITY = 32 * 2 ** 20  # 32 MB
+    _CURRENT_BUFFER_SIZE = 0
+
+    def __init__(self, filename, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self._filename = filename
+
+    @staticmethod
+    def _hash(blob):
+        """Calculate and return the md5 hash value for the file data.
+
+        Parameters
+        ----------
+        blob : bytes
+            Byte literal to be hashed.
+
+        Returns
+        -------
+        str
+            The md5 hash of the input bytes.
+        """
+        if blob is not None:
+            m = hashlib.md5()
+            m.update(blob)
+            return m.hexdigest()
+
+    def _get_file_metadata(self):
+        """Return metadata of file.
+
+        Returns
+        -------
+        Tuple[int, float] or None
+            The size and last modification time of the associated file. If the
+            file does not exist, returns :code:`None`.
+        """
+        try:
+            metadata = os.stat(self._filename)
+            return metadata.st_size, metadata.st_mtime
+        except OSError as error:
+            if error.errno != errno.ENOENT:
+                raise
+            else:
+                # A return value of None indicates that the file does not
+                # exist and can be reliably compared against for metadata-based
+                # validation checks.
+                return None
```

Comment:
No `else` needed. Also do you mean "canNOT" be compared for metadata validation?
```suggestion
            # A return value of None indicates that the file does not
            # exist and cannot be used for metadata-based validation.
            return None
```

Reply:
I meant that the `None` value can be reasonably compared against since `other_metadata == None` will reliably return `False`. I'll reword to clarify.